File: app/src/main/java/com/roman/gurdan/sudo/pro/util/TimerUtil.java
package com.roman.gurdan.sudo.pro.util;

public class TimerUtil {

    private long start = 0;
    private long duration = 0;
    private boolean canceled = false;

    public TimerUtil() {
    }

    public void start() {
        start = System.currentTimeMillis();
        this.canceled = false;
    }

    public void resume() {
        start = System.currentTimeMillis();
        this.canceled = false;
    }

    public void pause() {
        duration += System.currentTimeMillis() - start;
        this.canceled = true;
    }

    public void cancel() {
        this.pause();
    }

    public boolean isCanceled() {
        return canceled;
    }

    public long getDuration() {
        if (this.canceled) return duration;
        else
            return System.currentTimeMillis() - start + duration;
    }

}

File: app/src/main/java/com/roman/gurdan/sudo/pro/util/TimerUtil.kt
package com.roman.gurdan.sudo.pro.util

class TimerUtil {

    private var start = 0L
    private var duration = 0L
    private var canceled = false

    fun start() {
        start = System.currentTimeMillis()
        this.canceled = false
    }

    fun resume() {
        this.start()
    }

    fun pause() {
        duration += System.currentTimeMillis() - start
        this.canceled = true
    }

    fun cancel() {
        if (isCanceled()) return
        this.pause()
    }

    fun isCanceled(): Boolean = canceled

    fun reset() {
        duration = 0
        this.start()
    }

    fun getDuration(): Long {
        return if (isCanceled()) duration else System.currentTimeMillis() - start + duration
    }


}